util-vserver/man/reducecap.8
.TH "reducecap" "8" "0.1.0" "Klavs Klavsen <kl@vsen.dk>" "System Administration"
.SH "NAME"
.LP 
reducecap \- The reducecap utility is used to lower the capability ceiling of a process and child process.
.SH "SYNTAX"
.LP 
reducecap [\fIoptions\fP] <\fIcommand arguments\fP>
.SH "DESCRIPTION"
.LP 
The reducecap utility is used to lower the capability ceiling of a process and child process. Even setuid program won't be able to grab more capabilities.
.SH "OPTIONS"
.LP 
.TP 
\fB\-\-secure\fR Removes all dangerous capabilities from the process executed.Specificly it removes:
CAP_LINUX_IMMUTABLE CAP_NET_BROADCAST CAP_NET_ADMIN, CAP_NET_RAW CAP_IPC_LOCK CAP_IPC_OWNER CAP_SYS_MODULE CAP_SYS_RAWIO CAP_SYS_PACCT CAP_SYS_ADMIN CAP_SYS_BOOT CAP_SYS_NICE CAP_SYS_RESOURCE CAP_SYS_TIME CAP_MKNOD.

Leaving the following capabilities:
CAP_CHOWN CAP_DAC_OVERRIDE CAP_DAC_READ_SEARCH CAP_FOWNER CAP_FSETID CAP_KILL CAP_SETGID CAP_SETUID CAP_NET_BIND_SERVICE CAP_SYS_CHROOT CAP_SYS_PTRACE CAP_SYS_TTY_CONFIG CAP_LEASE CAP_OPENDEV
.TP 
\fB\-\-show\fR Shows the current process capabilities.
.TP 
\fB\-\-flag\fR
sets the security context flags. The option may be repeated
several times. Here are the values:

lock: The security context can't be changed. The process is trapped
	  in this context. This is generally used for vservers because yoy
	  do not want them to hide in new security context.

sched: Each process in a security context contribute (lower) to the general
	priority of every processes in the context. Mostly, all processes
	in a security context take as much CPU together as one process
	not bound to this flag. Said again differently, a vserver having
	100 active processes won't get more CPU than another vserver
	with a single active process.


nproc: The "ulimit -u N" setting becomes global to the security context. It means
	the security context is not allowed to have more than N processes.

private: No other processes, even root in security context 0, is allowed to
	enter this security context. Once a security context is setup
	with this flag, it is on its own. This also means that root
	in security context 0 won't be able to kill or interact with those
	processes.

hideinfo: Hides various information in /proc.

.TP 
\fB--LINUX_IMMUTABLE\fR
.TP 
\fB--NET_BIND_SERVICE\fR
.TP 
\fB--NET_BROADCAST\fR
.TP 
\fB--NET_ADMIN\fR
.TP 
\fB--NET_RAW\fR
.TP 
\fB--IPC_LOCK\fR
.TP 
\fB--IPC_OWNER\fR
.TP 
\fB--SYS_MODULE\fR
.TP 
\fB--SYS_RAWIO\fR
.TP 
\fB--SYS_PACCT\fR
.TP 
\fB--SYS_ADMIN\fR
.TP 
\fB--SYS_BOOT\fR
.TP 
\fB--SYS_NICE\fR
.TP 
\fB--SYS_RESOURCE\fR
.TP 
\fB--SYS_TIME\fR
.TP 
\fB--MKNOD\fR

All these options remove one capability. These options may be used
after the
\fB--secure\fR
option to remove more capabilities.


.SH "FILES"
.LP 
\fI/usr/sbin/reducecap\fP 


.SH "EXAMPLES"
.LP 
# You are not root now
# What is the current capability ceiling
cat /proc/self/status
# The capBset line presents mostly 1s.
/usr/sbin/reducecap \-\-secure /bin/sh
cat /proc/self/status
# The capBset now shows many more 0s.
# The capEff shows all 0s, you have no privilege now
# We su to root
su
cat /proc/self/status
# capEff is much better now, but there are still many 0s
# Now we try to see if we are really root
tail /var/log/messages
# So far so good, we see the content
/sbin/ifconfig eth0
/sbin/ifconfig eth0 down
# No way, we can't configure the interface. In fact
# we have lost most privilege normally assigned to root
exit
.LP 
Please contribute some more, if you feel it's important.
.SH "AUTHORS"
.LP 
This Man page was written by Klavs Klavsen <kl@vsen.dk> and based upon the helpful output from the program itself and the documentation on the Virtual Server site <http://www.solucorp.qc.ca/miscprj/s_context.hc?prjstate=1&nodoc=0>
.SH "SEE ALSO"
.LP 
chcontext(8) rebootmgr(8) chbind(8)
vps(8) vpstree(8) vrpm(8) vserver(8)
vserver\-stat(8) vtop(8)


NEWS
version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


util-vserver/src/reducecap.c
// $Id$

// Copyright (C) 2003 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
// based on reducecap.cc by Jacques Gelinas
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif
#include "compat.h"

#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#include "vserver.h"
#include "linuxcaps.h"

#ifndef CAP_QUOTACTL
#  define CAP_QUOTACTL		VC_CAP_QUOTACTL
#endif

extern int capget (struct __user_cap_header_struct *, struct __user_cap_data_struct *);
extern int capset (struct __user_cap_header_struct *, struct __user_cap_data_struct *);

static void usage()
{
	fprintf (stderr,"reducecap version %s\n",VERSION);
	fprintf (stderr,"reducecap [ options ] command argument\n");
	exit (-1);
}

static void reducecap_print(struct __user_cap_data_struct *user)
{
	int i;
	printf ("%22s %9s %9s %9s\n","Capability","Effective","Permitted"
		,"Inheritable");
	for (i=0;; ++i) {
	        char const *	text = vc_cap2text(i);
		int		bit  = 1<<i;
		if (text==0) break;
		
		printf ("%-22s %9s %9s %9s\n"
			,text
			,(user->effective   & bit) ? "X    " : " "
			,(user->permitted   & bit) ? "X    " : " "
			,(user->inheritable & bit) ? "X    " : " ");
	}
}

static void reducecap_show()
{
	struct __user_cap_header_struct header;
	struct __user_cap_data_struct user;
	header.version = _LINUX_CAPABILITY_VERSION;
	header.pid = getpid();
	if (capget(&header,&user)==-1){
		perror ("capget");
	}else{
		reducecap_print (&user);
	}
}



int main (int argc, char *argv[])
{
	int ret = -1;
	unsigned long remove = 0;
	int show = 0;
	int flags = 0;
	unsigned long secure = (1<<CAP_LINUX_IMMUTABLE)
		|(1<<CAP_NET_BROADCAST)
		|(1<<CAP_NET_ADMIN)
		|(1<<CAP_NET_RAW)
		|(1<<CAP_IPC_LOCK)
		|(1<<CAP_IPC_OWNER)
		|(1<<CAP_SYS_MODULE)
		|(1<<CAP_SYS_RAWIO)
		|(1<<CAP_SYS_PACCT)
		|(1<<CAP_SYS_ADMIN)
		|(1<<CAP_SYS_BOOT)
		|(1<<CAP_SYS_NICE)
		|(1<<CAP_SYS_RESOURCE)
		|(1<<CAP_SYS_TIME)
		|(1<<CAP_MKNOD)
	        |(1<<CAP_QUOTACTL);
	int i;
	for (i=1; i<argc; i++){
		const char *arg = argv[i];
		const char *opt = argv[i+1];
		if (strcmp(arg,"--secure")==0){
			remove = secure;
		}else if (strcmp(arg,"--show")==0){
			show = 1;
		}else if (strcmp(arg,"--flag")==0){
			if (strcmp(opt,"lock")==0){
				flags |= 1;
			}else if (strcmp(opt,"sched")==0){
				flags |= 2;
			}else if (strcmp(opt,"nproc")==0){
				flags |= 4;
			}else if (strcmp(opt,"private")==0){
				flags |= 8;
			}else if (strcmp(opt,"hideinfo")==0){
				flags |= 32;
			}else{
				fprintf (stderr,"Unknown flag %s\n",opt);
			}
			i++;
		}else if (arg[0] == '-' && arg[1] == '-'){
			static struct {
				const char *option;
				int bit;
			}tbcap[]={
				// The following capabilities are normally available
				// to vservers administrator, but are place for
				// completeness
				{"CAP_CHOWN",CAP_CHOWN},
				{"CAP_DAC_OVERRIDE",CAP_DAC_OVERRIDE},
				{"CAP_DAC_READ_SEARCH",CAP_DAC_READ_SEARCH},
				{"CAP_FOWNER",CAP_FOWNER},
				{"CAP_FSETID",CAP_FSETID},
				{"CAP_KILL",CAP_KILL},
				{"CAP_SETGID",CAP_SETGID},
				{"CAP_SETUID",CAP_SETUID},
				{"CAP_SETPCAP",CAP_SETPCAP},
				{"CAP_SYS_TTY_CONFIG",CAP_SYS_TTY_CONFIG},
				{"CAP_LEASE",CAP_LEASE},
				{"CAP_SYS_CHROOT",CAP_SYS_CHROOT},

				// Those capabilities are not normally available
				// to vservers because they are not needed and
				// may represent a security risk
				{"--LINUX_IMMUTABLE",CAP_LINUX_IMMUTABLE},
				{"--NET_BIND_SERVICE",CAP_NET_BIND_SERVICE},
				{"--NET_BROADCAST",CAP_NET_BROADCAST},
				{"--NET_ADMIN",	CAP_NET_ADMIN},
				{"--NET_RAW",	CAP_NET_RAW},
				{"--IPC_LOCK",	CAP_IPC_LOCK},
				{"--IPC_OWNER",	CAP_IPC_OWNER},
				{"--SYS_MODULE",CAP_SYS_MODULE},
				{"--SYS_RAWIO",	CAP_SYS_RAWIO},
				{"--SYS_PACCT",	CAP_SYS_PACCT},
				{"--SYS_ADMIN",	CAP_SYS_ADMIN},
				{"--SYS_BOOT",	CAP_SYS_BOOT},
				{"--SYS_NICE",	CAP_SYS_NICE},
				{"--SYS_RESOURCE",CAP_SYS_RESOURCE},
				{"--SYS_TIME",	CAP_SYS_TIME},
				{"--MKNOD",		CAP_MKNOD},
				{"--QUOTACTL",          CAP_QUOTACTL},
				{NULL,0}
			};
			int j;
			for (j=0; tbcap[j].option != NULL; j++){
				if (strcasecmp(tbcap[j].option,arg)==0){
					remove |= (1<<tbcap[j].bit);
					break;
				}
			}
			if (tbcap[j].option != NULL){
				usage();
			}
		}else{
			break;
		}
	}
	if (i == argc){
		if (show){
			reducecap_show();
		}else{
			usage();
		}
	}else if (argv[i][0] == '-'){
		usage();
	}else{
		struct __user_cap_header_struct header;
		struct __user_cap_data_struct user;
		header.version = _LINUX_CAPABILITY_VERSION;
		header.pid = 0;
		if (capget(&header,&user)==-1){
			perror ("capget");
		}else{
			if (show){
				reducecap_print (&user);
			}
			if (vc_new_s_context(VC_SAMECTX,remove,flags)==VC_NOCTX){
				perror ("new_s_context -2");
			}else{
				fprintf (stderr,"Executing\n");
				execvp (argv[i],argv+i);
				fprintf (stderr,"Can't execute command %s\n",argv[i]);
			}
		}
	}
	return ret;
}

